cosmwasm/contracts/wormchain-ibc-receiver/src/ibc.rs

```diff
@@ -0,0 +1,170 @@
+use cosmwasm_std::{
+    entry_point, from_slice, to_binary, Binary, ContractResult, DepsMut, Env,
+    Ibc3ChannelOpenResponse, IbcBasicResponse, IbcChannelCloseMsg, IbcChannelConnectMsg,
+    IbcChannelOpenMsg, IbcChannelOpenResponse, IbcPacketAckMsg, IbcPacketReceiveMsg,
+    IbcPacketTimeoutMsg, IbcReceiveResponse, Response, StdError, StdResult,
+};
+
+use crate::msg::WormholeIbcPacketMsg;
+
+// Implementation of IBC protocol
+// Implements 6 entry points that are required for the x/wasm runtime to bind a port for this contract
+// https://github.com/CosmWasm/cosmwasm/blob/main/IBC.md#writing-new-protocols
+
+pub const IBC_APP_VERSION: &str = "ibc-wormhole-v1";
+
+/// 1. Opening a channel. Step 1 of handshake. Combines ChanOpenInit and ChanOpenTry from the spec.
+/// The only valid action of the contract is to accept the channel or reject it.
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn ibc_channel_open(
+    _deps: DepsMut,
+    _env: Env,
+    msg: IbcChannelOpenMsg,
+) -> StdResult<IbcChannelOpenResponse> {
+    let channel = msg.channel();
+
+    if channel.version.as_str() != IBC_APP_VERSION {
+        return Err(StdError::generic_err(format!(
+            "Must set version to `{}`",
+            IBC_APP_VERSION
+        )));
+    }
+
+    if let Some(counter_version) = msg.counterparty_version() {
+        if counter_version != IBC_APP_VERSION {
+            return Err(StdError::generic_err(format!(
+                "Counterparty version must be `{}`",
+                IBC_APP_VERSION
+            )));
+        }
+    }
+
```

Comment:
Should we be checking CHAIN_CONNECTIONS somewhere in here?  Otherwise it doesn't look like that map is used anywhere.

Reply:
See the whitepaper PR for how the CHAIN_CONNECTIONS map should be used.

TLDR - it's a mapping that we need to maintain somewhere for security, so that the guardian watcher can verify where IBC messages come from. @bruce-riley @evan-gray @hendrikhofstadt and I decided to keep it in this ibc-wormchain-receiver contract, and only allow updates through governance VAAs.